src/error.rs

```diff
@@ -24,224 +27,153 @@
-            SacnParsePackError(sacn_parse_pack_error::Error, sacn_parse_pack_error::ErrorKind);
+    impl IntoSacnError for uuid::Error {
+        fn into_sacn_error(self, context: String) -> SacnError {
+            SacnError::UuidError(format!("{}: {}", context, self))
         }
+    }
+
+    impl IntoSacnError for std::io::Error {
+        fn into_sacn_error(self, context: String) -> SacnError {
+            SacnError::IoError(format!("{}: {}", context, self))
+        }
+    }
+
+    pub trait WithContext<T, E> {
+        fn with_context<F>(self, f: F) -> std::result::Result<T, SacnError>
+        where
+            F: FnOnce() -> String;
+    }
+
+    impl<T, E> WithContext<T, E> for std::result::Result<T, E>
+    where
+        E: IntoSacnError,
+    {
+        fn with_context<F>(self, f: F) -> std::result::Result<T, SacnError>
+        where
+            F: FnOnce() -> String,
+        {
+            self.map_err(|e| e.into_sacn_error(f()))
+        }
+    }
+
+    impl<T> WithContext<T, SacnError> for std::result::Result<T, SacnError> {
+        fn with_context<F>(self, f: F) -> std::result::Result<T, SacnError>
+        where
+            F: FnOnce() -> String,
+        {
+            self.map_err(|e| SacnError::Context {
+                context: f(),
+                source: Box::new(e),
+            })
+        }
+    }
+    
+
+    #[derive(Debug, Error)]
+    pub enum SacnError {
+        #[error("IO error")]
+        Io(#[from] io::Error),
+
+        #[error("UTF8 error")]
+        Utf8(#[from] Utf8Error),
+
+        #[error("UUID parse error")]
+        Uuid(#[from] uuid::Error),
+
+        #[error("UTF8 error: {0}")]
+        Utf8Error(String),
+
+        #[error("UUID error: {0}")]
+        UuidError(String),
+
+        #[error("IO error: {0}")]
+        IoError(String),
+
+        #[error(transparent)]
+        SacnParsePackError(#[from] SacnParsePackError),
+
+        #[error("The given source name was malformed and couldn't be used, msg: {0}")]
+        MalformedSourceName(String),
+
+        #[error("Attempted to perform an action using a priority value that is invalid, msg: {0}")]
+        InvalidPriority(String),
+
+        #[error("Limit for the number of supported sources has been reached, msg: {0}")]
+        SourcesExceededError(String),
+
+        #[error(
+            "A source was discovered by a receiver with the announce_discovery_flag set to true, source name: {0}"
+        )]
+        SourceDiscovered(String),
+
+        #[error("Attempted to exceed the capacity of a single universe, msg: {0}")]
+        ExceedUniverseCapacity(String),
+
+        #[error("Illegal universe used, outwith allowed range, msg: {0}")]
+        IllegalUniverse(String),
+
+        #[error("Attempted to use a universe that wasn't first registered for use, msg: {0}")]
+        UniverseNotRegistered(String),
+
+        #[error("IP version (ipv4 or ipv6) used when the other is expected, msg: {0}")]
+        IpVersionError(String),
```

Comment:
Same comment again about how we should remove the msg bit where possible as its not very computer usable